java-checks/src/main/java/org/sonar/java/checks/ThisExposedFromConstructorCheck.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2018 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import com.google.common.collect.ImmutableList;
+import java.util.List;
+import org.sonar.check.Rule;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.tree.Arguments;
+import org.sonar.plugins.java.api.tree.AssignmentExpressionTree;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+@Rule(key = "S3366")
+public class ThisExposedFromConstructorCheck extends IssuableSubscriptionVisitor {
+
+  private static final String REPORT = "Make sure the use of \"this\" doesn't" +
+    " expose partially-constructed instances of this class in multi-threaded environments.";
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return ImmutableList.of(Tree.Kind.CONSTRUCTOR);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (!hasSemantic()) {
+      return;
+    }
+    MethodTree methodTree = (MethodTree) tree;
+    methodTree.block().accept(new ConstructorBodyVisitor(methodTree.symbol().owner()));
+  }
+
+  private class ConstructorBodyVisitor extends BaseTreeVisitor {
+    private Symbol owner;
+
+    public ConstructorBodyVisitor(Symbol owner) {
+      this.owner = owner;
+    }
+
+    @Override
+    public void visitMethodInvocation(MethodInvocationTree tree) {
```

Comment:
you need to call `super.VisitMethodInvocation(tree)` at the end of this method, otherwise, you are going to cut the exploration of child, and therefore miss any nested method invocation in the parameters:

`foo(myObject.bar(this))`

Reply:
@stylianos-agapiou-sonarsource This comment was not addressed. You can add a failing testcase using the example provided in the previous comment. And then fix it following the suggestion of Michael.